editor/plugins/version_control_editor_plugin.cpp

```diff
@@ -97,219 +113,566 @@ void VersionControlEditorPlugin::popup_vcs_set_up_dialog(const Control *p_gui_ba
+		ProjectSettings::get_singleton()->set("version_control/plugin_name", selected_addon);
+		ProjectSettings::get_singleton()->save();
+	}
+}
+
+bool VersionControlEditorPlugin::_load_addon(String p_name) {
+	String path = ScriptServer::get_global_class_path(p_name);
 	Ref<Script> script = ResourceLoader::load(path);
 
-	ERR_FAIL_COND_MSG(!script.is_valid(), "VCS Addon path is invalid");
+	ERR_FAIL_COND_V_MSG(!script.is_valid(), false, "VCS Plugin path is invalid");
 
 	EditorVCSInterface *vcs_interface = memnew(EditorVCSInterface);
 	ScriptInstance *addon_script_instance = script->instance_create(vcs_interface);
 
-	ERR_FAIL_COND_MSG(!addon_script_instance, "Failed to create addon script instance.");
+	ERR_FAIL_COND_V_MSG(!addon_script_instance, false, "Failed to create plugin script instance.");
 
 	// The addon is attached as a script to the VCS interface as a proxy end-point
 	vcs_interface->set_script_and_instance(script.get_ref_ptr(), addon_script_instance);
 
 	EditorVCSInterface::set_singleton(vcs_interface);
+
+	return true;
+}
+
+void VersionControlEditorPlugin::_set_up() {
+	register_editor();
 	EditorFileSystem::get_singleton()->connect("filesystem_changed", this, "_refresh_stage_area");
 
 	String res_dir = OS::get_singleton()->get_resource_dir();
 
-	ERR_FAIL_COND_MSG(!EditorVCSInterface::get_singleton()->initialize(res_dir), "VCS was not initialized");
+	ERR_FAIL_COND_MSG(!EditorVCSInterface::get_singleton()->initialize(res_dir), "VCS was not initialized.");
 
 	_refresh_stage_area();
+	_refresh_commit_list();
+	_refresh_branch_list();
 }
 
-void VersionControlEditorPlugin::_send_commit_msg() {
-	if (EditorVCSInterface::get_singleton()) {
-		if (staged_files_count == 0) {
-			commit_status->set_text(TTR("No files added to stage"));
-			return;
-		}
+void VersionControlEditorPlugin::_set_credentials() {
+	if (set_up_username->get_text() == "" || set_up_password->get_text() == "") {
+		WARN_PRINT("Some features will not work without remote credentials.")
+	}
+
+	EditorVCSInterface::get_singleton()->set_up_credentials(set_up_username->get_text(), set_up_password->get_text());
+	set_up_username->clear();
+	set_up_password->clear();
+}
 
-		EditorVCSInterface::get_singleton()->commit(commit_message->get_text());
+void VersionControlEditorPlugin::_refresh_branch_list() {
+	CHECK_PLUGIN_INITIALIZED();
+	List<String> branch_list = EditorVCSInterface::get_singleton()->get_branch_list();
+	branch_select->clear();
+	for (int i = 0; i < branch_list.size(); i++) {
+		branch_select->add_item(branch_list[i], i);
+	}
 
-		commit_message->set_text("");
-		version_control_dock_button->set_pressed(false);
-	} else {
-		WARN_PRINT("No VCS addon is initialized. Select a Version Control Addon from Project menu");
+	// TODO: Add new brach feature.
+	// branch_select->add_separator();
+	// branch_select->add_item("New Branch");
+	branch_select->select(0); // First branch in the list is current branch
```

Comment:
We should ideally also check to see if it is a bare repository with 0 branches, but it is really rare to see a user create a bare repository because most Git repositories need to sync to a remote Git repo to actually function as a reliable VCS solution.